blockchain/utxocache.go

```diff
@@ -122,3 +129,406 @@ func (ms *mapSlice) deleteMaps() {
+// overwritten with the given output.  All fields will be updated for existing
+// entries since it's possible it has changed during a reorg.
+func (s *utxoCache) addTxOut(outpoint wire.OutPoint, txOut *wire.TxOut, isCoinBase bool, blockHeight int32) error {
+	// Don't add provably unspendable outputs.
+	if txscript.IsUnspendable(txOut.PkScript) {
+		return nil
+	}
+
+	entry := new(UtxoEntry)
+	entry.amount = txOut.Value
+	// Deep copy the script when the script in the entry differs from the one in
+	// the txout.  This is required since the txout script is a subslice of the
+	// overall contiguous buffer that the msg tx houses for all scripts within
+	// the tx.  It is deep copied here since this entry may be added to the utxo
+	// cache, and we don't want the utxo cache holding the entry to prevent all
+	// of the other tx scripts from getting garbage collected.
+	entry.pkScript = make([]byte, len(txOut.PkScript))
+	copy(entry.pkScript, txOut.PkScript)
+
+	entry.blockHeight = blockHeight
+	entry.packedFlags = tfFresh | tfModified
+	if isCoinBase {
+		entry.packedFlags |= tfCoinBase
+	}
+
+	s.cachedEntries.put(outpoint, entry, s.totalEntryMemory)
+	s.totalEntryMemory += entry.memoryUsage()
+
+	return nil
+}
+
+// addTxOuts adds all outputs in the passed transaction which are not provably
+// unspendable to the view.  When the view already has entries for any of the
+// outputs, they are simply marked unspent.  All fields will be updated for
+// existing entries since it's possible it has changed during a reorg.
+func (s *utxoCache) addTxOuts(tx *btcutil.Tx, blockHeight int32) error {
+	// Loop all of the transaction outputs and add those which are not
+	// provably unspendable.
+	isCoinBase := IsCoinBase(tx)
+	prevOut := wire.OutPoint{Hash: *tx.Hash()}
+	for txOutIdx, txOut := range tx.MsgTx().TxOut {
+		// Update existing entries.  All fields are updated because it's
+		// possible (although extremely unlikely) that the existing
+		// entry is being replaced by a different transaction with the
+		// same hash.  This is allowed so long as the previous
+		// transaction is fully spent.
+		prevOut.Index = uint32(txOutIdx)
+		err := s.addTxOut(prevOut, txOut, isCoinBase, blockHeight)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// addTxIn will add the given input to the cache if the previous outpoint the txin
+// is pointing to exists in the utxo set.  The utxo that is being spent by the input
+// will be marked as spent and if the utxo is fresh (meaning that the database on disk
+// never saw it), it will be removed from the cache.
+func (s *utxoCache) addTxIn(txIn *wire.TxIn, stxos *[]SpentTxOut) error {
+	// Ensure the referenced utxo exists in the view.  This should
+	// never happen unless there is a bug is introduced in the code.
+	entries, err := s.fetchEntries([]wire.OutPoint{txIn.PreviousOutPoint})
+	if err != nil {
+		return err
+	}
+	if len(entries) < 1 || entries[0] == nil {
```

Comment:
check `len(entries) != 1` instead?

Reply:
Changed to `len(entries) != 1 || entries[0] == nil` in the latest push